### PR TITLE
[Merged by Bors] - feat: `dupNamespace` as a syntax linter

### PR DESCRIPTION
**Pull request by adomani**
An implementation of the `dupNamespace` from `Std` as a syntax linter.  The linter emits a warning for declarations that have a repeated, consecutive namespace:
* `Nat.Nat.foo` triggers the linter;
* `Nat.foo.Nat` does *not* trigger the linter.

*Note.* This linter will not detect duplication in namespaces of autogenerated declarations (other than the one whose `declId` is present in the source declaration).

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60dupNamespace.60.20linter)

The linter found
```lean
private theorem Seminorm.Seminorm.isLUB_sSup ...
```
that is skipped by the environment `dupNamespace` linter.

---

A note on the extra import: it is only used to get `Array.attach`, which allows to write
```lean
def getIds : Syntax → Array Syntax
  | stx@(.node _ _ args) =>
    ((args.attach.map fun ⟨a, _⟩ => getIds a).foldl (· ++ ·) #[stx]).filter (·.getKind == ``declId)
  | _ => default
```
and not have to mark it as `partial`.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
!bench

**Review by YaelDillies — APPROVED** [review]
maintainer merge

File: test/Simps.lean [hunk 1]
```diff
@@ -123,6 +123,7 @@ example (n : ℕ) : foo.rfl.toFun n = n := by rw [foo.rfl_toFun, id]
 example (n : ℕ) : foo.rfl.invFun n = n := by rw [foo.rfl_invFun]
 
 /- the declarations are `simp` lemmas -/
+set_option linter.dupNamespace false in
```
Comment (mattrobball) [inline test/Simps.lean:126]: Why not change `foo` to `bar` and make the linter happy?

**Comment by mattrobball** [comment]
It would be nice to have tests for the behavior under exporting and aliasing also.

**Comment by adomani** [comment]
!bench

File: test/Simps.lean (hunk 1 above)
Comment (adomani) [reply test/Simps.lean]: Thanks for the suggestion: I changed the names and removed the `set_option`.

**Comment by adomani** [comment]
I added checking for `alias`: it was slightly awkward, since the "name" node for `alias` is `ident`, while the one for `theorem/lemma` is `declId`.  I have not checked what happens with `export`, though.

**Comment by adomani** [comment]
> Here are the [benchmark results](http://speed.lean-fro.org/mathlib4/run-detail/8b702ad7-6e6d-4803-9989-b3a19d4d4935) for commit [9c29121](https://github.com/leanprover-community/mathlib4/commit/9c291216088fbaf85afc731e301beb0685fe03e9). The entire run failed. Found no significant differences.

The bench should happenat #11747.

**Comment by mattrobball** [comment]
Great!

Can you write a test for naming via export and comment it out with some explanation if it doesn't work?

bors delegate+

**Comment by adomani** [comment]
Matt, thanks!  I added support for `export` and a test.  Assuming that CI is happy with the change, I will merge!

**Pull request closed without merging**